src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/StackWalk/FrameHandling/FrameIterator.cs

```diff
@@ -0,0 +1,160 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.Diagnostics.DataContractReader.Contracts.StackWalkHelpers;
+
+internal sealed class FrameIterator
+{
+    private enum FrameType
+    {
+        Unknown,
+
+        InlinedCallFrame,
+        SoftwareExceptionFrame,
+
+        /* TransitionFrame Types */
+        FramedMethodFrame,
+        CLRToCOMMethodFrame,
+        PInvokeCalliFrame,
+        PrestubMethodFrame,
+        StubDispatchFrame,
+        CallCountingHelperFrame,
+        ExternalMethodFrame,
+        DynamicHelperFrame,
+
+        FuncEvalFrame,
+
+        /* ResumableFrame Types */
+        ResumableFrame,
+        RedirectedThreadFrame,
+
+        FaultingExceptionFrame,
+
+        HijackFrame,
+    }
+
+    private readonly Target target;
+    private readonly TargetPointer terminator;
+    private TargetPointer currentFramePointer;
+
+    internal Data.Frame CurrentFrame => target.ProcessedData.GetOrAdd<Data.Frame>(currentFramePointer);
+
+    public TargetPointer CurrentFrameAddress => currentFramePointer;
+
+    public FrameIterator(Target target, ThreadData threadData)
+    {
+        this.target = target;
+        terminator = new TargetPointer(target.PointerSize == 8 ? ulong.MaxValue : uint.MaxValue);
+        currentFramePointer = threadData.Frame;
+    }
+
+    public bool IsValid()
+    {
+        return currentFramePointer != terminator;
+    }
+
+    public bool Next()
+    {
+        if (currentFramePointer == terminator)
+            return false;
+
+        currentFramePointer = CurrentFrame.Next;
+        return true;
```

Comment:
Should this return false if CurrentFrame.Next was the terminator, like IEnumerator.MoveNext does when you reach the end of a sequence?